iocontrol/raspi/controller.go

```diff
@@ -1,51 +1,138 @@
 package raspi
 
-import "context"
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"periph.io/x/conn/v3/gpio"
+	"periph.io/x/conn/v3/gpio/gpioreg"
+	"periph.io/x/host/v3"
+)
+
+const (
+	_loggerName = "raspi_controller"
+)
+
+// Pins reference by board number (1-40)
+// CAN HAT uses 15, 19, 21, 23, 24, 26, 33
+// HAT EEPROM uses 27, 28
+// UART 8, 10 and I2C 3, 5 are optional
+var boardToBCM = map[uint8]int{
+	7: 4, 11: 17, 12: 18, 13: 27, 16: 23, 18: 24, 22: 25,
+	29: 5, 31: 6, 32: 12, 35: 19, 36: 16, 37: 26, 38: 20, 40: 21,
+}
 
 // Controller provides control for various Raspberry Pi pins
 type Controller struct {
+	l      *zap.Logger
+	mu     sync.Mutex
+	opened bool
 }
 
 // NewController returns a new Raspberry Pi controller
-func NewController() *Controller {
-	return &Controller{}
+func NewController(l *zap.Logger) *Controller {
+	return &Controller{
+		l: l.Named(_loggerName),
+	}
 }
 
 // Open configures the controller
 func (c *Controller) Open(_ context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.opened {
+		return errors.New("raspberry pi controller already opened")
+	}
+
+	// initialize periph io
+	if _, err := host.Init(); err != nil {
+		return errors.Wrap(err, "periph host init")
+	}
+
+	c.opened = true
 	return nil
 }
 
-// SetDigital sets an output digital pin for a Raspberry Pi digital pin
-func (c *Controller) SetDigital(output *DigitalPin, b bool) error {
+// Closes the controller instance
```

Comment:
nit: usually we start the comment for the exposed function with the name of the function 
```suggestion
// Close tears down the controller instance
```